SYCL/ESIMD/tpm_tests.cpp

```diff
@@ -0,0 +1,244 @@
+//==--------------- tpm_tests.cpp - DPC++ ESIMD on-device test -------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// TODO enable on Windows and Level Zero
+// REQUIRES: linux && gpu && opencl
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 1
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 2
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out 3
+
+// Since in ESIMD a single WI occupies entire underlying H/W thread, SYCL
+// private memory maps to what's known as 'thread private memory' in CM.
+// This test is intended to use TPM to support implementation in ESIMD
+// backend. In order to force using of TPM need to allocate 96x32 bytes or more.
+
+#include "esimd_test_utils.hpp"
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+#include <iostream>
+
+using namespace cl::sycl;
+
+int main(int argc, char **argv) {
+  constexpr unsigned VL = 8;
+  constexpr unsigned SZ = 800; // big enough to use TPM
+
+  if (argc != 2) {
+    std::cout << "Skipped! Specify case number" << std::endl;
+    return 1;
+  }
+
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+  auto ctx = q.get_context();
+
+  int *output =
+      static_cast<int *>(sycl::malloc_shared(VL * sizeof(int), dev, ctx));
+  memset(output, 0, VL * sizeof(int));
+
+  int case_num = atoi(argv[1]);
+  std::cout << "CASE NUM: " << case_num << std::endl;
+
+  int offx1 = 111;
+  int offx2 = 55;
+  int offy1 = 499;
+  int offy2 = 223;
+  int offz = 99;
+  int base1 = 500;
+  int base2 = 100;
+  int divisor = 4;
+
+  {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Test>(
+          sycl::range<1>{1}, [=](id<1> i) SYCL_ESIMD_KERNEL {
+            using namespace sycl::INTEL::gpu;
+            simd<int, VL> val(0);
+
+            int x1[SZ];
+            for (int j = 0; j < SZ; ++j) {
```

Comment:
can some of this code be outlined into a function, which can be reused on the host side as well?